pulsar-client-cpp/lib/ConsumerImpl.cc

```diff
@@ -1228,8 +1243,15 @@ void ConsumerImpl::seekAsync(const MessageId& msgId, ResultCallback callback) {
         uint64_t requestId = client->newRequestId();
         LOG_DEBUG(getName() << " Sending seek Command for Consumer - " << getConsumerId() << ", requestId - "
                             << requestId);
+        auto chunkMessageIdImpl = std::dynamic_pointer_cast<ChunkMessageIdImpl>(msgId.impl_);
         Future<Result, ResponseData> future =
-            cnx->sendRequestWithId(Commands::newSeek(consumerId_, requestId, msgId), requestId);
+            (chunkMessageIdImpl != nullptr)
+                ? cnx->sendRequestWithId(
+                      Commands::newSeek(consumerId_, requestId,
+                                        chunkMessageIdImpl->getFirstChunkMessageIdImpl().ledgerId_,
+                                        chunkMessageIdImpl->getFirstChunkMessageIdImpl().entryId_),
+                      requestId)
+                : cnx->sendRequestWithId(Commands::newSeek(consumerId_, requestId, msgId), requestId);
```

Comment:
I think we can eliminate the following overload of `newSeek`.

```c++
SharedBuffer newSeek(uint64_t consumerId, uint64_t requestId, const MessageId& messageId);
```

Just get the ledger id and entry id here and use the newly added overload.

```c++
        const auto ledgerId = (chunkMessageIdImpl ? chunkMessageIdImpl->getFirstChunkMessageIdImpl().ledgerId_
                                                  : msgId.ledgerId());
        const auto entryId = (chunkMessageIdImpl ? chunkMessageIdImpl->getFirstChunkMessageIdImpl().ledgerId_
                                                 : msgId.entryId());
        Future<Result, ResponseData> future =
            cnx->sendRequestWithId(Commands::newSeek(consumerId_, requestId, ledgerId, entryId), requestId);
```

----

But I think a better solution is checking the type of `MessageIdImpl` in `Commands::newSeek` like:

```c++
    const auto msgIdImpl = messageId.impl_;
    const auto chunkMessageIdImpl = std::dynamic_pointer_cast<ChunkMessageIdImpl>(msgIdImpl);
    if (chunkMessageIdImpl) {
        const auto& firstMsgId = chunkMessageIdImpl->getFirstChunkMessageIdImpl();
        messageIdData.set_ledgerid(firstMsgId.ledgerId_);
        messageIdData.set_entryid(firstMsgId.entryId_);
    } else {
        messageIdData.set_ledgerid(msgIdImpl->ledgerId_);
        messageIdData.set_entryid(msgIdImpl->entryId_);
    }
```

Reply:
I also think check in newSeek is better.

Reply:
I think is better to save the overload
`newSeek(uint64_t consumerId, uint64_t requestId, const MessageId& messageId);`
add check impl in it, choose ledgerId and entryId then return
`newSeek(uint64_t consumerId, uint64_t requestId, int64_t ledgerId, int64_t entryId)`
what do you think?

Reply:
I save the overload
`newSeek(uint64_t consumerId, uint64_t requestId, const MessageId& messageId);`
add check impl in it, choose ledgerId and entryId then return.
So the `newSeek(uint64_t consumerId, uint64_t requestId, int64_t ledgerId, int64_t entryId)` I will delete it.